components/brave_rewards/renderer/creator_detection_agent.cc

```diff
@@ -0,0 +1,265 @@
+  std::string source;
+  bool should_run_in_isolated_world = true;
+};
+
+std::optional<DetectionScript> GetDetectionScript(
+    content::RenderFrame* render_frame) {
+  // Only run scripts for the main frame.
+  if (!render_frame || !render_frame->IsMainFrame()) {
+    return std::nullopt;
+  }
+
+  // Only run scripts if the user has enabled Brave Rewards.
+  if (!render_frame->GetBlinkPreferences().brave_rewards_enabled) {
+    return std::nullopt;
+  }
+
+  auto* web_frame = render_frame->GetWebFrame();
+  DCHECK(web_frame);
+  if (web_frame->IsProvisional()) {
+    return std::nullopt;
+  }
+
+  // Only run scripts for known "media platform" sites.
+  GURL origin_url = url::Origin(web_frame->GetSecurityOrigin()).GetURL();
+  if (!IsMediaPlatformURL(origin_url)) {
+    return std::nullopt;
+  }
+
+  // Only run scripts when there is an exact hostname match.
+  auto iter = kScriptMap.find(origin_url.host_piece());
+  if (iter == kScriptMap.end()) {
+    return std::nullopt;
+  }
+
+  return DetectionScript{
+      .source = blink::Platform::Current()->GetDataResourceString(iter->second),
+      .should_run_in_isolated_world = ShouldRunInIsolatedWorld(iter->second)};
+}
+
+v8::MaybeLocal<v8::Function> CompileDetectionInitializer(
+    v8::Local<v8::Context> context,
+    std::string_view function_body) {
+  v8::Context::Scope context_scope(context);
+  v8::MicrotasksScope microtasks_scope(
+      context, v8::MicrotasksScope::kDoNotRunMicrotasks);
+
+  auto* isolate = context->GetIsolate();
+  v8::ScriptCompiler::Source source(gin::StringToV8(isolate, function_body));
+
+  v8::Local<v8::String> parameters[] = {
+      gin::StringToV8(isolate, "setPageChangedCallback"),
+      gin::StringToV8(isolate, "onCreatorDetected"),
+      gin::StringToV8(isolate, "verboseLogging")};
+
+  // The creator detection script is compiled as a function body. Browser
+  // capabilities are passed in to the compiled function as arguments whose
+  // parameter names are defined above. Browser capabilities are only exposed
+  // to the script; they are not visible outside of the compiled function.
+  return v8::ScriptCompiler::CompileFunction(context, &source, 3, parameters);
+}
+
+template <typename... Args>
+void CallFunctionWithArgs(v8::Isolate* isolate,
+                          v8::Local<v8::Function> function,
+                          Args... args) {
+  v8::Local<v8::Context> context = function->GetCreationContextChecked(isolate);
+  v8::Context::Scope context_scope(context);
+  v8::MicrotasksScope microtasks_scope(
+      context, v8::MicrotasksScope::kDoNotRunMicrotasks);
+  std::array<v8::Local<v8::Value>, sizeof...(Args)> arg_list = {
+      args.template As<v8::Value>()...};
+  std::ignore = function->Call(context, context->Global(), arg_list.size(),
+                               arg_list.empty() ? nullptr : arg_list.data());
+}
+
+}  // namespace
+
+CreatorDetectionAgent::CreatorDetectionAgent(content::RenderFrame* render_frame,
```

Comment:
It's not clear to me why this needs to happen on the renderer. Why can't we just execute this from a tab helper or similar?

Reply:
@bridiver we need not just execute, but also get a result and updates back multiple times.

Reply:
we can get a delayed result back using a promise from the browser process. Are you saying there are cases where we might get more than one result on the same page load? Does it change or could there be more than one on a page?